### ISSUE TITLE
blurHandler with datePicker, sets back the $viewValue to the mask.

### ISSUE THREAD

**Issue by Rouche** [angular-ui/ui-mask#157]
I have a mask 9999-99-99 for a date. (combined with a bootstrap-ui datepicker)

1) Lets say i enter a valid date and blur out. At the end of the blurHandler() i will have:
prevValue = "20100101"
$viewValue = "2010-01-01"

2) Now i backspace the field EXACTLY on the last character, Not one more. blur out:
at the start of blurHandler()
prevValue = "20100101"
$viewValue = ""
3) triggerChangeEvent(iElement[0]); happens

$viewValue is set to "-  -", wich is invalid. And i have no idea where this value come from.

**Well it comes from the Event trigger, the "value" is trimmed in angular listener and setViewValue() is called.** 

4) If i do the same thing again, but i backspace ONCE more than the exact number of character, it triggers some sort of patch wich does this thing and seems to correct the problem.
```
scope.$apply(function () {
    controller.$setViewValue(''); // $setViewValue should be run in angular context, otherwise the changes will be invisible to angular and user code.
});
```

Im trying to find a workaround, but it seems you already had this problem.

Any idea how i could keep the viewValue to its real value ( "" ).

**Comment by Rouche** [comment]
I got an idea. Do you think this code could do the job in blurHandler()
```
if (value !== prevValue) {
    triggerChangeEvent(iElement[0]);
    //Fix $viewValue becoming invalid.
    controller.$viewValue = value;
}
```

**Comment by lukepfeiffer10** [comment]
Can you provide a plunkr that shows this issue? I'm having trouble recreating the issue you have described.

**Comment by Rouche** [comment]
Probably be tomorrow.

**Comment by Rouche** [comment]
Plnkr: http://plnkr.co/edit/FlQH8l3cuNTKHMP0bs2G?p=preview

**Comment by lukepfeiffer10** [comment]
So, I realized that this is only happening when the placeholder character is `space` which is because angular most likely only sets the value if `trim` actually modifies the string and this would only happen in the case of a `space`.

I'm going to keep looking into this as it does have some weird behavior. However, I am curious if this is actually causing any problems. Could you elaborate on what problem if any you are experiencing; other than the `$viewValue` doesn't appear to be correct.

**Comment by Rouche** [comment]
In fact the case is complex.

It have to do with the $valid flags and validations errors.

In this particular case, the flag ng-invalid and ng-invalid-date too. But ng-valid-mask is ok.

The problem is theres no way to determine between an ng-invalid-date + ng-valid-mask when the date entered is : 2016-99-99, except the $viewValue being empty.

**Comment by Rouche** [comment]
Updated plunk with $isEmpty(). Wich seems not in sync with $viewValue.

**Comment by lukepfeiffer10** [comment]
Yeah I found that yesterday too and it's because of the `space` mask placeholder being used that is causing the discrepancy (sort of). If you don't have a `space` for the mask placeholder character then when it gets to this line in the angular source code:

```
if (ctrl.$viewValue !== value || (value === '' && ctrl.$$hasNativeValidators)) {
      ctrl.$setViewValue(value, event);
}
```

It ends up triggering the ui-mask `parser` function which then sets the view value again back to what you would expect. However, when using a `space` for the placeholder character the `parser` function is never being called... 

It's a rather tricky bug and I understand at least why your fix is working now I just wish I could find another way of solving the issue.

**Comment by lukepfeiffer10** [comment]
After digging into the code some more the true culprit is this section of code in the angular source:

```
/**
   * @ngdoc method
   * @name ngModel.NgModelController#$commitViewValue
   *
   * @description
   * Commit a pending update to the `$modelValue`.
   *
   * Updates may be pending by a debounced event or because the input is waiting for a some future
   * event defined in `ng-model-options`. this method is rarely needed as `NgModelController`
   * usually handles calling this in response to input events.
   */
  this.$commitViewValue = function() {
    var viewValue = ctrl.$viewValue;

    $timeout.cancel(pendingDebounce);

    // If the view value has not changed then we should just exit, except in the case where there is
    // a native validator on the element. In this case the validation state may have changed even though
    // the viewValue has stayed empty.
    if (ctrl.$$lastCommittedViewValue === viewValue && (viewValue !== '' || !ctrl.$$hasNativeValidators)) {
      return;
    }
    ctrl.$$updateEmptyClasses(viewValue);
    ctrl.$$lastCommittedViewValue = viewValue;

    // change to dirty
    if (ctrl.$pristine) {
      this.$setDirty();
    }
    this.$$parseAndValidate();
  };
```

The main problem is still that the placeholder character is a `space` but what that causes to happen is the `$$lastCommittedViewValue === viewValue` in the above `if` check. This means that it will never hit the `this.$$parseAndValidate()` which means it will never get back to the ui-mask `parser` function and allow that to fix the `$viewValue`.

The reason `$$lastCommittedViewValue` and `viewValue` are the same is because angular always trims the elements value and you are using `space` as the placeholder character which, when deleting characters from your masked input, all that you are really doing is replacing the characters you entered with a `space`. This being an issue when angular trims the value because any number of spaces gets removed meaning angular will think your view value hasn't changed even though we know it has changed.


TL:DR; spaces as a mask placeholder are bad when angular goes to process the element.

I think a better fix in this situation might be to make a check for this case in the `blurHandler` and call `$setViewValue`. I think you were going to do this once?

Like so:

```
//Check for different value and trigger change.
if (value !== prevValue) {
    triggerChangeEvent(iElement[0]);
    if (value === '' && angular.isDefined(uiMaskPlaceholderChar) && uiMaskPlaceholderChar === 'space') {
        scope.$apply(function() {
            controller.$setViewValue('');
        });
    }
}
```

**Comment by Rouche** [comment]
Yep thats what i did first, but it causes a flicker of values. The triggered one, and then the new one set.

This makes the validators run for nothing, $flag some values that will disapear in next digest.

**Comment by lukepfeiffer10** [comment]
That's too bad about the flickering... Have you noticed it being a problem any other time then when `value === ''`? Just want to limit this as much as possible because this also seems like it would cause the element to flicker since you are setting the element's value and then resetting it afterward.

**Comment by Rouche** [comment]
No just when value is ```''```. I am unable to see a value flick tough. Maybe i should try on a slower computer. But it is very minimal because its only the fill character that might disapear for a millisecond. More acceptable than an error message :)

I understand about the space. I would have kept the underscores myself. But i do not choose the design. And rightfully so hehe :P

**Comment by lukepfeiffer10** [comment]
Designers always making it harder for the developer... :P 

Okay well if you wouldn't mind updating your PR then I'll get it merged in later today.

**Comment by lukepfeiffer10** [comment]
Released in 1.8.5

**Comment by Rouche** [comment]
Sweet, thanks!
